etcdserver/server.go

```diff
@@ -1024,8 +1024,37 @@ func (s *EtcdServer) parseProposeCtxErr(err error, start time.Time) error {
 		if start.After(prevLeadLost) && start.Before(curLeadElected) {
 			return ErrTimeoutDueToLeaderFail
 		}
+
+		lead := types.ID(atomic.LoadUint64(&s.r.lead))
+		switch lead {
+		case types.ID(raft.None):
+			// TODO: return error to specify it happens because the cluster does not have leader now
+		case s.ID():
+			if s.isQuorumLosed(start) {
+				return ErrTimeoutDueToConnectionLost
+			}
+		default:
+			if t := s.r.transport.ActiveSince(lead); t.IsZero() || t.After(start) {
+				return ErrTimeoutDueToConnectionLost
+			}
+		}
+
 		return ErrTimeout
 	default:
 		return err
 	}
 }
+
+func (s *EtcdServer) isQuorumLosed(check time.Time) bool {
```

Comment:
Can we change this to IsConnectedToQuorumAt(at time.Time) bool and add a comment to it. Ideally, this function does not need to be a method on etcdserver